### DISCUSSION TITLE
Add a stopwatch to help user track time spent on certain habits

### DISCUSSION THREAD

**Discussion by ghost** [iSoron/uhabits#90] [Feature Requests]
Hello, thanks for this app. It's really useful :)

It would be nice to be able to set a (optional) timer when we define a habit.
The timer would be a chronometer (for habits like read at least 30 minutes a day) or a countdown (don't play videogames more than x)
At the begining of the day 'reading' should be uncrossed, 'playing' crossed. Each time we do the habit we start/unpause the timer (a notification for the timer would be great), when we are done we pause it. When the timer reaches the limit (30 or 0) it crosses or uncrosses the habit.

**Comment by iSoron** [comment]
This is an interesting suggestion, but I would prefer to consider it after we add support to numerical habits #42. If the user has a habit such as 'how many hours/minutes did you work on this today?', then the app could provide a timer that would increment the number automatically.

**Comment by cyberscientist** [comment]
Is this feature going in? Because I think this would be quite a useful feature.

**Reply by hiqua** [reply]
Numerical habits are in v2, which is currently in alpha.

**Comment by ke352802081770314** [comment]
I seem to be lucky to have started with Loop just after v2 was released. 
I noticed that the "numerical" is entirely generic, and the "unit" is specified by the user as free text, therefore clearly doesn't affect the handling by the app.
I think the app would bebefit from breaking down numeric in three alternatives:
* count (unitless): would give the option "+1" instead of edit the number, so that if something was done multiple times a day it can simply count up.
* time spent: I am not sure whether minutes is good enough for everyone or whether we might benefit from permitting seconds. Assuming minutes is granular enough, the "count" could be interpreted as minutes, and maybe displayed as h:mm if more than 59. In that case, clicking on it could pop up a menu including "start" "stop" which manages a clock that counts the duration spent on that habit item, or "enter duration" to record it afterwards. So all I would have to do is click "start" when I start jogging/walking/gardening/whatever and "stop" when finishing instead of having to track the time by some other means so that I know what to enter into Loop after completion. (currently I am just guessing it which may be far off the truth.) Maybe, if the stopwatch is currently running (i.e. I have pressed start and not yet stop), the app should display a "stop doing xxx" dialogue prominently instead of the usual UI or as a notification so that the clock can't accidentally be kept running.
* anything else: as it is now, see also the "add to current value" feature I just responded to.

**Reply by hiqua** [reply]
I agree that timed habits could benefit from some specific handling. For "count" I think it's fine to let the user specify "times".

**Comment by alextio** [comment]
Has there been any progress on this front?

**Comment by KraXen72** [comment]
this would indeed be pretty useful. right now i just use my phone stopwatch, but there are a few frictions points:
- i have to switch between apps to track time (clock <=> loop habits)
- since the stopwatch counts in min/hour/sec, and loop habit having general units (min), i have to always convert hours+min+sec into min. approximation works fine most of the time & the math isn't hard, but after i've been doing a habit for let's say 1h26min (86min) and do it for another 3 hours and 17minutes, you can see it's a slight friction point.
  
 a time habit style, where you have a very simple stopwatch/countdown that preserves state across day, and can be accessed through notification (if set) would be really useful.
 => for example, my habit to practise guitar would notify me at 8:00am, ill hit the start button in notification, practise for x minutes, hit stop and it would remember i pracised for x minutes. if x > habit goal (after stopping stopwatch), notif goes away by itself (stopwatch still accessible in app if practising even more, if not, notification still stays (the 2nd button in the notification is "dismiss notification" => give user option to not do the habit that day basically, same as if you'd click no on a boolean habit.
  
 tldr not a big deal, atleast i am forced to do basic math lol, but it would be useful.
